src/pip/_internal/cache.py

```diff
@@ -15,21 +17,68 @@
 from pip._internal.models.direct_url import DirectUrl
 from pip._internal.models.link import Link
 from pip._internal.models.wheel import Wheel
+from pip._internal.req.req_install import InstallRequirement
 from pip._internal.utils.temp_dir import TempDirectory, tempdir_kinds
 from pip._internal.utils.urls import path_to_url
+from pip._internal.vcs import vcs
 
 logger = logging.getLogger(__name__)
 
+_egg_info_re = re.compile(r"([a-z0-9_.]+)-([a-z0-9_.!+-]+)", re.IGNORECASE)
+
 ORIGIN_JSON_NAME = "origin.json"
 
 
+def _contains_egg_info(s: str) -> bool:
+    """Determine whether the string looks like an egg_info.
+
+    :param s: The string to parse. E.g. foo-2.1
+    """
+    return bool(_egg_info_re.search(s))
+
+
+def should_cache(
+    req: InstallRequirement,
+) -> bool:
+    """
+    Return whether a built InstallRequirement can be stored in the persistent
+    wheel cache, assuming the wheel cache is available, and _should_build()
+    has determined a wheel needs to be built.
+    """
+    if not req.link:
+        return False
+
+    if req.editable or (not req.source_dir and not req.link.is_wheel):
+        # never cache editable requirements
+        return False
```

Comment:
The first part of the conditional covers editable, what does the latter part do?

Reply:
This is actually one of the remaining TODOs I needed to figure out. When I modified this PR to use the `should_cache()` method, I didn't realize that the `should_cache()` method assumed that `req.source_dir` had already been populated, so I made the change you highlighted here so that wheel dists would *also* be entered into the `LinkMetadataCache`.

I think this is fixed in the current version of the PR by avoiding the use of the `should_cache()` method when *looking up* cached metadata, and only calling `should_cache()` to determine whether to populate the `LinkMetadataCache`. This gets us the result @sbidoul described: only populating the `LinkMetadataCache` for sdists, while leaning on the other forms of metadata-only dists for other cases.